nptdms/writer.py

```diff
@@ -30,29 +59,63 @@ def __init__(self, file, mode='w', version=4712):
             It's important that if you are appending segments to an
             existing TDMS file, this matches the existing file version (this can be queried with the
             :py:attr:`~nptdms.TdmsFile.tdms_version` property).
+        :param index_file: Whether or not to write a index file besides the data file. Index files
+            can be used to accelerate reading speeds for faster channel extraction and data positions inside
+            the data files. Only valid if submitted file variable is a path.
         """
         valid_versions = (4712, 4713)
         if version not in valid_versions:
             raise ValueError("version must be one of %s" % ",".join("%d" % v for v in valid_versions))
         self._file = None
+        self._index_file = None
         self._file_path = None
+        self._index_file_path = None
         self._file_mode = mode
         self._tdms_version = version
 
         if hasattr(file, "read"):
             # Is a file
             self._file = file
+            if isinstance(index_file, BytesIO):
+                self._index_file = index_file
+            elif isinstance(index_file, bool) and not index_file:
+                pass
+            else:
+                raise ValueError(
+                    f"Invalid type, ``index_file`` can only be ``False`` or a stream to write into, "
+                    "but is {type(index_file)}"
+                )
         else:
             self._file_path = file
+            if isinstance(index_file, bool):
+                if index_file:
+                    self._index_file_path = file + "_index"
+            else:
+                raise ValueError(
+                    f"Invalid type, ``index_file`` can  only be ``False`` or ``True`` but is {type(index_file)}."
+                )
 
     def open(self):
         if self._file_path is not None:
             self._file = open(self._file_path, self._file_mode + 'b')
+            if self._index_file_path is not None:
+                self._index_file = open(self._index_file_path, self._file_mode + 'b')
 
     def close(self):
-        if self._file_path is not None:
+        if self._file_path is None:
+            if self._file is not None:
+                self._file.seek(0, os.SEEK_SET)
```

Comment:
I don't think we need the behaviour of seeking to the beginning of the file when closing, and this could be a breaking change for anyone expecting to be able to continue writing to the end of a stream.